java/vector/src/main/java/org/apache/arrow/vector/dictionary/ListSubfieldEncoder.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.dictionary;
+
+import org.apache.arrow.vector.BaseIntVector;
+import org.apache.arrow.vector.BitVectorHelper;
+import org.apache.arrow.vector.ValueVector;
+import org.apache.arrow.vector.complex.BaseRepeatedValueVector;
+import org.apache.arrow.vector.complex.FixedSizeListVector;
+import org.apache.arrow.vector.complex.ListVector;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.util.TransferPair;
+
+/**
+ * Sub fields encoder/decoder for Dictionary encoded {@link ListVector} and {@link FixedSizeListVector}.
+ */
+public class ListSubfieldEncoder {
+
+  static ValueVector encodeFixedSizeListSubField(ValueVector v, Dictionary dictionary) {
+    FixedSizeListVector vector = (FixedSizeListVector) v;
+    FixedSizeListVector dictVector = (FixedSizeListVector) dictionary.getVector();
+    final int listSize = vector.getListSize();
+
+    DictionaryHashTable hashTable = new DictionaryHashTable(dictVector.getDataVector());
+    for (int i = 0; i < dictVector.getDataVector().getValueCount(); i++) {
+      hashTable.put(i);
+    }
+
+    Field valueField = vector.getField();
+    FieldType indexFieldType = new FieldType(valueField.isNullable(), dictionary.getEncoding().getIndexType(),
+        dictionary.getEncoding(), valueField.getMetadata());
+
+    FixedSizeListVector encoded = FixedSizeListVector.empty(vector.getName(), listSize, vector.getAllocator());
+    encoded.allocateNewSafe();
+    encoded.setValueCount(vector.getValueCount());
+    BaseIntVector indices = (BaseIntVector) encoded.addOrGetVector(indexFieldType).getVector();
+
+
+    for (int i = 0; i < v.getValueCount(); i++) {
+      if (!vector.isNull(i)) {
+        int start = i * listSize;
+        int end = (i + 1) * listSize;
+        ValueVector dataVector = vector.getDataVector();
+
+        BitVectorHelper.setValidityBitToOne(encoded.getValidityBuffer(), i);
+        for (int k = start; k < end; k++) {
```

Comment:
doesn't this duplicate logic from the normal encode?

Reply:
Yes, I think could extra a method for reuse here